### PR TITLE
Issue 874 - Implement a ComposerArtifactCoordinates

### LINKED ISSUE #874: Implement a ComposerArtifactCoordinates
We need to identify what the artifact coordinates are for Composer packages and implement a class for this. This task requires someone with basic PHP and Composer knowledge.

The code should be located in a class called `ComposerArtifactCoordinates` under `strongbox/strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-composer-layout-provider/src/main/java/org/carlspring/strongbox/artifact/coordinates`.

Have a look at the [Artifact Coordinates](https://github.com/strongbox/strongbox/wiki/Artifact-Coordinates) wiki page for more information.

This is a sub-task of #872. It should be relatively straight-forward to implement this and you should get acquainted with the master task, before proceeding.

### PR DESCRIPTION
**Pull request by jcoelho93**
Closes #874

### CODE REVIEW

**Comment by carlspring** [comment]
Hello again!

Thank you for you contribution and congratulations on your first pull request!

Just as a reminder being carried over from the other pull request that we closed. 
> Would you mind signing the ICLA, as described in the [Contributing](https://github.com/strongbox/strongbox/blob/master/CONTRIBUTING.md) page?

:)

**Comment by carlspring** [comment]
Some notes... Could you please implement:
* A method for parsing the coordinates from a path?
* A method that returns the path from coordinates?
* A test case that checks the above two methods using real names of Composer package?

Perhaps you'lll find [this recent pull request](https://github.com/strongbox/strongbox/pull/1197/files#diff-875171b08b711c972a4eba9ed6f860c3R201) useful. We need something similar.

**Comment by steve-todorov** [comment]
Hi @jcoelho93 and thank you for your contribution! :) 

I've checked your PR and it look good to me. In addition to @carlspring's notes, could you also please add the `type` property as well?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-composer-layout-provider/src/test/java/org/carlspring/strongbox/artifact/coordinates/ComposerArtifactCoordinatesTest.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    static
+    {
+        packages = new ArrayList<>();
+    }
+
+    @BeforeAll
+    static void setup()
+    {
+
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-composer-layout-provider/src/test/java/org/carlspring/strongbox/artifact/coordinates/ComposerArtifactCoordinatesTest.java]: Cosmetic: Too many empty lines all over. It makes sens to put an empty line before and after a logical block to make it stand out, but in this case, it looks like it's a bit too much. Would you mind removing these (here and in the other method)? Thanks! :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-composer-layout-provider/src/test/java/org/carlspring/strongbox/artifact/coordinates/ComposerArtifactCoordinatesTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+        packages.add("symfony/polyfill-mbstring");
+
+        packages.add("psr/log");
+
+        packages.add("symfony/polyfill-ctype");
+
+        packages.add("guzzlehttp/psr7");
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-composer-layout-provider/src/test/java/org/carlspring/strongbox/artifact/coordinates/ComposerArtifactCoordinatesTest.java]: Would you mind adding some more test data (or even more tests) checking cases with, or without versions, types and so on? Thanks! :)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-composer-layout-provider/src/test/java/org/carlspring/strongbox/artifact/coordinates/ComposerArtifactCoordinatesTest.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    }
+
+    @Test
+    public void testCreateArtifactFromParse()
+    {
+
+        for (String _package : packages)
```
Comment (carlspring) [inline strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-composer-layout-provider/src/test/java/org/carlspring/strongbox/artifact/coordinates/ComposerArtifactCoordinatesTest.java]: Cosmetic: Could you please change the variable's name to `package`, or `p`?

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-composer-layout-provider/src/test/java/org/carlspring/strongbox/artifact/coordinates/ComposerArtifactCoordinatesTest.java (hunk 2 above)
Comment (jcoelho93) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-composer-layout-provider/src/test/java/org/carlspring/strongbox/artifact/coordinates/ComposerArtifactCoordinatesTest.java]: Hey @carlspring,
 How would you test with version and types?
In Composer, the path does not include version and type. There is not a representation like with Maven `symfony:polyfill:1.2.3`.

Thanks.
;)

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-composer-layout-provider/src/test/java/org/carlspring/strongbox/artifact/coordinates/ComposerArtifactCoordinatesTest.java (hunk 3 above)
Comment (jcoelho93) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-composer-layout-provider/src/test/java/org/carlspring/strongbox/artifact/coordinates/ComposerArtifactCoordinatesTest.java]: Done!

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-composer-layout-provider/src/test/java/org/carlspring/strongbox/artifact/coordinates/ComposerArtifactCoordinatesTest.java (hunk 1 above)
Comment (jcoelho93) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-composer-layout-provider/src/test/java/org/carlspring/strongbox/artifact/coordinates/ComposerArtifactCoordinatesTest.java]: Fixed

File: strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-composer-layout-provider/src/test/java/org/carlspring/strongbox/artifact/coordinates/ComposerArtifactCoordinatesTest.java (hunk 2 above)
Comment (jcoelho93) [reply strongbox-storage/strongbox-storage-layout-providers/strongbox-storage-composer-layout-provider/src/test/java/org/carlspring/strongbox/artifact/coordinates/ComposerArtifactCoordinatesTest.java]: I have used the syntax `vendor/name/version/type`

**Comment by steve-todorov** [comment]
Hi @jcoelho93,

Sorry for the late reply - I haven't had a chance to dig deep into composer's source code to see how it requests packages. This is why I haven't replied yet. Overall the PR looks good to me. I hope to have some time soon so we can move this forward. Apologize for the delay! :)

**Pull request closed without merging**